diff --git a/plat/imx/imx8mm/imx8mm_bl31_setup.c b/plat/imx/imx8mm/imx8mm_bl31_setup.c
index aa1c336..d9b353d 100644
--- a/plat/imx/imx8mm/imx8mm_bl31_setup.c
+++ b/plat/imx/imx8mm/imx8mm_bl31_setup.c
@@ -310,7 +310,10 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 
 	/* Assign M4 to domain 1 */
 	mmio_write_32(IMX_RDC_BASE + 0x204, 0x1);
+#if 0
+	/* UART4, uncomment if M4 only access desired */
 	mmio_write_32(IMX_RDC_BASE + 0x518, 0xfc);
+#endif
 	mmio_write_32(IMX_RDC_BASE + 0x5A4, 0xf3);
 
 #if defined (CSU_RDC_TEST)
